src/Microsoft.AspNetCore.Mvc.Core/Internal/MiddlewareFilterBuilderService.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    /// <summary>
+    /// Builds a middleware pipeline after receiving the pipeline from a pipeline provider
+    /// </summary>
+    public class MiddlewareFilterBuilderService
+    {
+        private readonly ConcurrentDictionary<Type, Lazy<RequestDelegate>> _pipelinesCache
+            = new ConcurrentDictionary<Type, Lazy<RequestDelegate>>();
+        private readonly MiddlewareFilterConfigurationProvider _configurationProvider;
+
+        public IApplicationBuilder ApplicationBuilder { get; set; }
+
+        public MiddlewareFilterBuilderService(MiddlewareFilterConfigurationProvider middlewareFilterConfigurationProvider)
+        {
+            _configurationProvider = middlewareFilterConfigurationProvider;
+        }
+
+        public RequestDelegate GetPipeline(Type middlewarePipelineProviderType)
+        {
+            // Build the pipeline only once. This is similar to how middlewares registered in Startup are constructed.
+
+            var requestDelegate = _pipelinesCache.GetOrAdd(
+                middlewarePipelineProviderType,
+                key => new Lazy<RequestDelegate>(() => BuildPipeline(key)));
```

Comment:
Except it prevents the pipeline from being built more than once by parallel request, correct?